File: treversal/cpp-solution/boundary treversal.cpp
// addLeftNodes = O(H)
// addRightNodes = O(H)
// addLeafNodes = O(N)
// so overall time complexity is O(N) and space: O(H)[right subtree will take space] Note: not considering stack space in recursion call
class Solution {
public:
    bool isLeafNode(Node * root){
        if (root->left == nullptr && root->right == nullptr){
            return true;
        }
        return false;
    }
    void addLeftNodes(Node*root, vector<int> &result){
        Node * current = root->left;
        while (current){
            if (!isLeafNode(current)){
                result.push_back(current->data);
            }
            if (current->left) current = current->left;
            else current = current->right;
        }
    }
    void addLeafNodes(Node * root, vector<int> &result){
        if (isLeafNode(root)){
            result.push_back(root->data);
            return;
        }
        if (root->left) addLeafNodes(root->left, result);
        if (root->right) addLeafNodes(root->right, result);
    }
    void addRightNodes(Node * root, vector<int> &result){
        Node * current = root->right;
        vector<int> temp;
        while (current){
            if (!isLeafNode(current)) temp.push_back(current->data);
            if (current->right) current = current->right;
            else current = current->left;
        }
        while(temp.size() > 0){
            result.push_back(temp.back());
            temp.pop_back();
        }
    }
    vector <int> boundary(Node *root)
    {
        //Your code here
        vector<int> result;
        if (root == nullptr){
            return result;
        }
        if (!isLeafNode(root)) result.push_back(root->data);
        addLeftNodes(root, result);
        addLeafNodes(root, result);
        addRightNodes(root, result);
        return result;
    }
};

File: treversal/python solution/boundary treversal.py
# // addLeftNodes = O(H)
# // addRightNodes = O(H)
# // addLeafNodes = O(N)
# // so overall time complexity is O(N) and space: O(H)[right subtree will take space] Note: not considering stack space in recursion call
class Solution:
    def isLeafNode(self, root):
        if not root.left and not root.right:
            return True
        return False
    def addLeftBoundary(self, root, result):
      # root node is already proccessed initially
      # trevese the left subtree excluding leaf node in the path
        current = root.left
        while current:
            if not self.isLeafNode(current):
                result.append(current.data)
            if current.left:
                current = current.left
            else:
                current = current.right
    def addLeaves(self, root, result):
       # treverse all the leaf node
        if self.isLeafNode(root):
            result.append(root.data)
            return
        if root.left: self.addLeaves(root.left, result)
        if root.right: self.addLeaves(root.right, result)
    
    def addRightBoundary(self, root, result):
      # root node is already proccessed initially
      # treverse the right subtree excluding leaf node in the path
        current = root.right
        temp = []
        while current:
            if not self.isLeafNode(current):
                temp.append(current.data)
            if current.right:
                current = current.right
            else:
                current = current.left
        while temp:
            result.append(temp.pop())
    def printBoundaryView(self, root):
        # Code here
        result = []
        if root is None:
            return result
            
        if not self.isLeafNode(root):
            result.append(root.data);
        self.addLeftBoundary(root, result)
        self.addLeaves(root, result)
        self.addRightBoundary(root, result)
        return result
